testutils/util/artifact.py

```diff
@@ -0,0 +1,206 @@
+import io
+import os
+import random
+import tarfile
+import hashlib
+import json
+
+
+class Artifact:
+    """
+    Artifact provides a very simplistic implementation of mender artifact
+    that allows creating simple buffered artifact file objects that
+    resides in memory.
+    """
+
+    _dummySHA = "%064d" % 0
+
+    def __init__(self, artifact_name, device_types, payload=None):
+        """
+        :param artifact_name: name of the artifact (str)
+        :param device_types:  list of compatible device types (list)
+        :param payload:       optional payload to initialize the payload
+                              section (file/io.IOBase, str, bytes)
+        """
+        if not isinstance(device_types, list):
+            raise ValueError("device_types must be a list of strings")
+
+        self._filenames = ["version", "header.tar.gz"]
+        self._payloads = {}
+        self._provides = {"header-info": {"artifact_name": artifact_name}}
+        self._depends = {"header-info": {"device_type": device_types}}
+        self._payload_types = {}
+        self._shasums = {}
+
+        if payload is not None:
+            self.add_payload(payload)
+
+    def add_payload(self, fd, payload_type="rootfs_image", depends=None, provides=None):
+        """
+        add_payload adds another payload to the payload section.
+        :param fd:           "file descriptor" contains the payload
+                             (io.IOBase/file, str, bytes)
+        :param payload_type: type of payload contained in fd (str)
+        :param depends:      optional depends for this payload (dict)
+        :param provides:     optional provides for this payload (dict)
+        """
+        if isinstance(fd, str):
+            fd = io.BytesIO(fd.encode())
+        elif isinstance(fd, bytes):
+            fd = io.BytesIO(fd)
+        elif not isinstance(fd, io.IOBase):
+            raise ValueError("fd must be an instance of either io.FileIO, str or bytes")
+        filename = "data/%04d/%s" % (
+            len(self._payloads),
+            getattr(fd, "name", "rootfs-%04d.ext4" % random.randint(0, 10000)),
+        )
+
+        self._filenames.append(filename)
+        self._payloads[filename] = fd
+        self._payload_types[filename] = payload_type
+        if isinstance(depends, dict):
+            self._depends[filename] = depends
+        if isinstance(provides, dict):
+            self._provides[filename] = provides
```

Comment:
It they are not `None`, I would assert that they are dict instances, to make sure we generate an error if a user passes a list or something else.

Reply:
:+1: